Documentation/postmortems/v3.5-data-inconsistency.md

```diff
@@ -0,0 +1,138 @@
+Reproduction run etcd under high stress and randomly killed one of the members using SIGKILL signal (not recoverable immediate process death). 
+
+## Detection
+
+For single member cluster it is totally undetectable. 
+There is no mechanism or tool for verifying that state database matches WAL.  
+
+In cluster with multiple members it would mean that one of the members that crashed, will missing changes from failed apply workflow.
+This means that it will have different state of database and will return different hash via `HashKV` grpc call.
+
+There is an automatic mechanism to detect data inconsistency. 
+It can be executed during etcd start via `--experimental-initial-corrupt-check` and periodically via `--experimental-corrupt-check-time`.
+Both checks however have a flaw, they depend on `HashKV` grpc method, which might fail causing the check to pass.
+
+In multi member etcd cluster, each member can run with different performance and be at different stage of applying the WAL log.
+Comparing database hashes between multiple etcd members requires all hashes to be calculated at the same change.
+This is done by requesting hash for the same `revision` (version of key value store). 
+However, it will not work if the provided revision is not available on the members.
+This can happen on very slow members, or in cases where corruption has lead revision numbers to diverge.
+
+This means that for this issue, the corrupt check is only reliable during etcd start just after etcd crashes.
+
+## Impact
+
+We are not aware any cases of users reporting a data corruption in production environment.
+
+However, issue was critical enough to motivate a public statement. 
+Main impact comes from loosing user trust into etcd reliability.
+
+## Lessons learned
+
+### What went well
+
+* Multiple maintainers were able to work effectively on reproducing and fixing the issue. As they are in different timezones, there was always someone working on the issue.
+
+### What went wrong
+
+* No users enable data corruption detection as it is still an experimental feature introduced in v3.3. All reported cases where detected manually making it almost impossible to reproduce.
+* Etcd has functional tests designed to detect such problems, however they are unmaintained, flaky and are missing crucial scenarios.
+* Etcd v3.5 release was not qualified as well as previous ones. Older maintainers run manual qualification process that is no longer known or executed.
+* Etcd apply code is so complicated that fixing the data inconsistency took almost 2 weeks and multiple tries (). Fix needed to be so complicated that we needed to develop automatic validation for it (https://github.com/etcd-io/etcd/pull/13885).
+* When fixing the main data inconsistency we have found multiple other edge cases that could lead to data corruption (https://github.com/etcd-io/etcd/issues/13514, https://github.com/etcd-io/etcd/issues/13922, https://github.com/etcd-io/etcd/issues/13937).
+
```

Comment:
+ Not clearly went wrong... but the problem is that as OSS community, we don't have insight what's the production adoption of different etcd versions thus their maturity. We make version as PROD-ready after some internal feadback... to get diverse usage, but the user's hold on till someone else will discover issues. I have no more recommendation here than transparency between maintainers.

Reply:
I think lack of feedback loop here is a persistent issue. Without a feedback loop about feature usage we cannot make a good production recommendation.  

What do you think we can do to improve the situation? Could we maybe each out to CNCF about collecting usage data from users?

Reply:
I think we can probably start with collecting version and uptime information with opt-out by default. I know the Kubernetes community tried to do similar things. Do you know how well it goes?